File: schemas/compliance-grc/convex/compliance_taggables.ts
// compliance_taggables: Polymorphic join table linking tags to any compliance entity.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const complianceTaggables = defineTable({
  tagId: v.id("compliance_tags"),
  taggableType: v.union(
    v.literal("control"),
    v.literal("risk"),
    v.literal("policy"),
    v.literal("audit"),
    v.literal("finding"),
    v.literal("evidence")
  ),
  taggableId: v.string(),
})
  .index("by_tag_id_and_taggable_type_and_taggable_id", ["tagId", "taggableType", "taggableId"])
  .index("by_taggable_type", ["taggableType"])
  .index("by_taggable_id", ["taggableId"]);

File: schemas/compliance-grc/spacetimedb/compliance_taggables.rs
// compliance_taggables: Polymorphic join table linking tags to compliance entities.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum TaggableType {
    Control,  // type: String
    Risk,
    Policy,
    Audit,
    Finding,
    Evidence,
}

#[spacetimedb::table(name = compliance_taggables, public)]
pub struct ComplianceTaggable {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub tag_id: String, // UUID — FK → compliance_tags.id (cascade delete)
    // Composite unique: (tag_id, taggable_type, taggable_id)

    #[index(btree)]
    pub taggable_type: TaggableType,

    #[index(btree)]
    pub taggable_id: String, // UUID — polymorphic, no FK constraint

    pub created_at: Timestamp,
}
